accepted/2025/supporting-bearer-auth.md

```diff
@@ -0,0 +1,69 @@
+# Supporting Bearer Token Auth for Credential Providers
+
+- Jonatan Gonzalez ([jgonz120](https://github.com/jgonz120))
+- Issue [#12877](https://github.com/NuGet/Home/issues/12877)
+
+## Summary
+
+Update the NuGet CLI to allow for bearer token authentication when publishing packages.
+
+## Motivation
+
+NuGet.org wants to shift to using OIDC authentication for publishing packages, enabling the use of short lived, 15-minute, API keys for authentication.
+This would reduce the security risk of a key leak, since currently the long-lived keys can be valid for up to a year.
+
+Currently NuGet.org takes in the API key through the [X-NuGet-ApiKey](https://learn.microsoft.com/en-us/nuget/api/package-publish-resource#request-parameters) header.
+The CLI for pushing provides a warning if the ApiKey is not specified causing feeds, [like Azure Artifacts](https://learn.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops#publish-packages), to tell users to pass a placeholder value into the command.
+
+Additionally due to the [limitation in .NET](https://github.com/dotnet/runtime/issues/91867#issue-1889923702) we can’t send these types of credentials using the HttpClientHandler.
+To circumvent this Azure Artifacts sends their key using basic auth, passing the key as a password in a network credential.
+By working around the .NET limitation for bearer token usage, we can allow a more technically correct solution and allow other feeds to implement a credential provider that returns a bearer token.
+
+## Explanation
+
+### Functional explanation
+
+The push command will no longer display a warning when no API key is specified.
+
+When the NuGet client receives an unauthorized response, it will check to see if the response header includes bearer, if it does it’ll retry the request passing the credential as a bearer token in the authentication header.
+
+### Technical explanation
+
+[HttpSourceAuthenticationHandler.SendAsync](https://github.com/NuGet/NuGet.Client/blob/313eecd3af442ee2eeed2e6decf310858934ab21/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpSourceAuthenticationHandler.cs?plain=1#L68) will be updated to check if the previous response contains a WWW-Authenticate: Bearer header.
+If it does, then before sending the request the method will attempt to retrieve the token from the credentials provided and pass them in the authentication header if it’s available.
+
+[GetAuthenticationCredentialsRequest](https://github.com/NuGet/NuGet.Client/blob/ab6d3b886b32e9b8f68a7f1b299c43b0c72487dc/src/NuGet.Core/NuGet.Protocol/Plugins/Messages/GetAuthenticationCredentialsRequest.cs?plain=1#L12C25-L12C60) will be updated to include a new field, AcceptedHeaders, that will tell the credential providers which auth headers can be used.
```

Comment:
The name could be `WwwAuthenticate` to align with the response header and match the naming convention in .NET
https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.httpresponseheaders.wwwauthenticate?view=net-9.0

Ideally client does minimal parsing and passes the value verbatim to the credential provider (perhaps as a string or array of strings, akin to https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.headerstringvalues?view=net-9.0). It's the cred provider that knows how to interpret the auth scheme, realm, other auth parameters, etc.

Reply:
Maybe `WwwAuthenticateHeader` to be clear in context?